cidc_api/models/models.py

```diff
@@ -1025,9 +1091,92 @@ def get_by_object_url(object_url: str, session: Session):
         """
         return session.query(DownloadableFiles).filter_by(object_url=object_url).one()
 
+    @classmethod
+    @with_default_session
+    def list_object_urls(
+        cls, ids: List[int], session: Session, filter_: Callable[[Query], Query]
+    ) -> List[str]:
+        """Get all object_urls for a batch of downloadable file record IDs"""
+        query = session.query(cls.object_url).filter(cls.id.in_(ids))
+        query = filter_(query)
+        return [r[0] for r in query.all()]
+
+    @classmethod
+    def build_file_bundle_query(cls) -> Query:
+        """
+        Build a query that selects nested file bundles from the downloadable files table.
+        The `file_bundles` query below should produce one bundle per unique `trial_id` that
+        appears in the downloadable files table. Each bundle will have shape like:
+        ```
+          {
+              <type 1>: {
+                <purpose 1>: [<file id 1>, <file id 2>, ...],
+                <purpose 2>: [...]
+              },
+              <type 2>: {...}
+          }
+        ```
+        where "type" is something like `"Olink"` or `"Participants Info"` and "purpose" is a `FilePurpose` string.
+        """
+        id_bundles = (
+            select(
+                [
+                    cls.trial_id,
+                    cls.consolidated_upload_type.label("type"),
+                    cls.file_purpose.label("purpose"),
+                    func.json_agg(cls.id).label("ids"),
+                ]
+            )
+            .group_by(cls.trial_id, cls.consolidated_upload_type, cls.file_purpose)
+            .alias("id_bundles")
+        )
+        purpose_bundles = (
+            select(
+                [
+                    literal_column("id_bundles.trial_id"),
+                    literal_column("id_bundles.type"),
+                    func.json_object_agg(
+                        func.coalesce(
+                            literal_column("id_bundles.purpose"), "miscellaneous"
```

Comment:
I think we have something like `[no purpose]` as misc option already (in db records?)

Reply:
We don't actually. We have a value like `[no facet group]` in certain file records `facet_group` field, but file purposes are a) are purely views not materialized in the database and b) only have four "allowed" values (not strictly enforced), `analysis`, `source`, `clinical`, and `miscellaneous`, based on this https://github.com/CIMAC-CIDC/cidc-ngs-pipeline-api/blob/2c7887a0a312dd4ec87fbe85a2741242977b38ea/output_API.schema.json#L28